FhirToDataLake/src/Microsoft.Health.AnalyticsConnector.Core/Jobs/FhirToDataLakeProcessingJobSpliter.cs

```diff
@@ -0,0 +1,322 @@
+        private const string CountApiParameter = "count";
+        private const string TotalCountFieldKey = "total";
+        private const string LastUpdatedApiParameter = "_lastUpdated";
+        private const string LastUpdatedApiParameterDesc = "-_lastUpdated";
+        private const int TimeoutSecondsGetResourceCount = 90;
+
+        public FhirToDataLakeProcessingJobSpliter(
+            IApiDataClient dataClient,
+            IDiagnosticLogger diagnosticLogger,
+            ILogger<FhirToDataLakeProcessingJobSpliter> logger)
+        {
+            _dataClient = EnsureArg.IsNotNull(dataClient, nameof(dataClient));
+            _diagnosticLogger = EnsureArg.IsNotNull(diagnosticLogger, nameof(diagnosticLogger));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public int LowBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.LowBoundOfProcessingJobResourceCount;
+
+        public int HighBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.HighBoundOfProcessingJobResourceCount;
+
+        public async IAsyncEnumerable<SubJobInfo> SplitJobAsync(string resourceType, DateTimeOffset? startTime, DateTimeOffset endTime, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var totalCount = await GetResourceCountAsync(resourceType, startTime, endTime, cancellationToken);
+
+            if (totalCount <= HighBoundOfProcessingJobResourceCount)
+            {
+                // Return job with total count lower than high bound.
+                _logger.LogInformation($"Split one sub job with {totalCount} resources from {startTime} to {endTime} for resource type {resourceType}.");
+
+                yield return new SubJobInfo
+                {
+                    ResourceType = resourceType,
+                    TimeRange = new TimeRange() { DataStartTime = startTime, DataEndTime = endTime },
+                    ResourceCount = totalCount,
+                };
+            }
+            else
+            {
+                // Split large size job.
+                var splittingStartTime = DateTimeOffset.UtcNow;
+                _logger.LogInformation($"Start splitting {resourceType} job, total {totalCount} resources.");
+
+                SortedDictionary<DateTimeOffset, int> anchorList = await InitializeAnchorListAsync(resourceType, startTime, endTime, totalCount, cancellationToken);
+                var lastTimeStamp = anchorList.LastOrDefault().Key;
+
+                _logger.LogInformation($"Splitting {resourceType} job. Use {(DateTimeOffset.UtcNow - splittingStartTime).TotalMilliseconds} milliseconds to initilize anchor list.");
+                var lastSplitTimestamp = DateTimeOffset.UtcNow;
+                DateTimeOffset? nextJobEnd = null;
+                var jobCount = 0;
+
+                while (nextJobEnd == null || nextJobEnd < endTime)
```

Comment:
nit: multi timestamps used here, it is a little confused. maybe add comments for it.

Reply:
Refined the naming strategy here, thanks.